File: backend/app/models/asset.py
from datetime import datetime
from enum import Enum
from typing import Any, Dict, Optional

from sqlalchemy import ForeignKey
from sqlmodel import JSON, Column, Field, SQLModel

from app.core.utils import utc_now


class AssetType(str, Enum):
    URL = "url"
    FORM = "form"
    XHR = "xhr"


class AssetSource(str, Enum):
    HTML = "html"
    JS = "js"
    NETWORK = "network"
    DOM = "dom"
    ROBOTS_TXT = "robots_txt"
    SITEMAP = "sitemap"


class AssetBase(SQLModel):
    target_id: int = Field(
        sa_column=Column(ForeignKey("targets.id", ondelete="CASCADE"), nullable=False)
    )
    content_hash: str = Field(index=True, unique=True, max_length=64)
    type: AssetType = Field(default=AssetType.URL)
    source: AssetSource = Field(default=AssetSource.HTML)
    method: str = Field(default="GET", max_length=10)
    url: str = Field()
    path: str = Field()

    # JSONB fields
    request_spec: Optional[Dict[str, Any]] = Field(default=None, sa_column=Column(JSON))
    response_spec: Optional[Dict[str, Any]] = Field(
        default=None, sa_column=Column(JSON)
    )
    parameters: Optional[Dict[str, Any]] = Field(default=None, sa_column=Column(JSON))

    last_task_id: Optional[int] = Field(default=None)


class Asset(AssetBase, table=True):
    __tablename__ = "assets"

    id: Optional[int] = Field(default=None, primary_key=True)
    first_seen_at: datetime = Field(default_factory=utc_now)
    last_seen_at: datetime = Field(default_factory=utc_now)


class AssetDiscovery(SQLModel, table=True):
    __tablename__ = "asset_discoveries"

    id: Optional[int] = Field(default=None, primary_key=True)
    task_id: int = Field(
        sa_column=Column(ForeignKey("tasks.id", ondelete="CASCADE"), nullable=False)
    )
    asset_id: int = Field(
        sa_column=Column(ForeignKey("assets.id", ondelete="CASCADE"), nullable=False)
    )
    parent_asset_id: Optional[int] = Field(
        default=None, sa_column=Column(ForeignKey("assets.id", ondelete="CASCADE"))
    )
    discovered_at: datetime = Field(default_factory=utc_now)


class AssetRead(AssetBase):
    id: int
    first_seen_at: datetime
    last_seen_at: datetime

File: frontend/src/types/asset.ts
/**
 * Asset Types and Interfaces
 * Matches backend AssetRead schema exactly
 */

/**
 * Asset type enum - represents the type of attack surface
 */
export enum AssetType {
  /** URL endpoint */
  URL = 'url',
  /** HTML Form */
  FORM = 'form',
  /** XHR/Fetch API request */
  XHR = 'xhr'
}

/**
 * Asset source enum - where the asset was discovered
 */
export enum AssetSource {
  /** Found in HTML */
  HTML = 'html',
  /** Found in JavaScript */
  JS = 'js',
  /** Found via network monitoring */
  NETWORK = 'network',
  /** Found in DOM */
  DOM = 'dom'
}

/**
 * Asset entity returned from the backend API
 * Represents a discovered attack surface (URL, Form, or XHR endpoint)
 */
export interface Asset {
  /** Unique identifier */
  id: number;
  /** Target ID this asset belongs to */
  target_id: number;
  /** SHA256 content hash for deduplication (max 64 chars) */
  content_hash: string;
  /** Type of attack surface */
  type: AssetType;
  /** Discovery source */
  source: AssetSource;
  /** HTTP method (e.g., GET, POST) */
  method: string;
  /** Full URL (max 2048 chars) */
  url: string;
  /** URL path (max 2048 chars) */
  path: string;
  /** Request specification (JSONB) - headers, body, cookies */
  request_spec: Record<string, unknown> | null;
  /** Response specification (JSONB) - status, headers, body */
  response_spec: Record<string, unknown> | null;
  /** Parsed parameters (JSONB) - name, type, location, value */
  parameters: Record<string, unknown> | null;
  /** ID of the last task that discovered this asset */
  last_task_id: number | null;
  /** ISO 8601 datetime when first discovered */
  first_seen_at: string;
  /** ISO 8601 datetime when last seen */
  last_seen_at: string;
}
